kitty: use makeBinaryWrapper to wrap kitty executable

liaowang11 opened the pull request:
The default `wrapProgram` no longer works on macos 15.1, `makeBinaryWrapper` is recommended in the manual to wrap executables on darwin.

The error reported for troubleshooting: 
`_LSOpenURLsWithCompletionHandler() failed with error -54.`

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

XYenon commented:
Looks like this don't work.

<img width="800" alt="image" src="https://github.com/user-attachments/assets/77d5a979-6f9a-4132-9526-7a3888948723">

Scrumplex commented on pkgs/applications/terminal-emulators/kitty/default.nix:
```diff
@@ -85,6 +86,7 @@ buildPythonApplication rec {
     go_1_23
     fontconfig
   ] ++ lib.optionals stdenv.hostPlatform.isDarwin [
+    makeBinaryWrapper
```
IMO you can do this for Linux, too, as `makeBinaryWrapper` will generally reduce startup time

liaowang11 commented:
> Looks like this don't work.
> <img alt="image" width="800" src="https://private-user-images.githubusercontent.com/20698483/382221044-77d5a979-6f9a-4132-9526-7a3888948723.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA0NjEwODAsIm5iZiI6MTczMDQ2MDc4MCwicGF0aCI6Ii8yMDY5ODQ4My8zODIyMjEwNDQtNzdkNWE5NzktNmY5YS00MTMyLTk1MjYtN2EzODg4OTQ4NzIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAxVDExMzMwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzZGU5ZWVmYzJjODE2MzZjYzUxNDQwNmRiMzM5YzBmYWQ1NWFiODNmMTA5YmVkNTYzNzk0OTBkNmM5NDhkMmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9QX3jJqjGA58mG04Vp6uT8gUd8e6znbotVBu17QNrlo">

Sorry, just tested on mac m1, I will try on my intel mac when i get home.

liaowang11 committed:
kitty: Binary wrap kitty, and sign on darwin

Using shell to wrap is slow and not recommended, this alsos fixes issue
on aarch64-darwin sequoia 15.1.
Also auto sign kitty to prevent permission error on x86_64-darwin.

yzx9 commented:
4 packages built:
kitty kitty.kitten kitty.shell_integration kitty.terminfo

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 352795`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>kitty</li>
    <li>kitty.kitten</li>
    <li>kitty.shell_integration</li>
    <li>kitty.terminfo</li>
  </ul>
</details>


## Results

Before:

```
› open ~/Applications/Home\ Manager\ Apps/kitty.app
_LSOpenURLsWithCompletionHandler() failed with error -54.
```

Now (This PR):

<img width="556" alt="image" src="https://github.com/user-attachments/assets/7b19b885-239a-4f0e-b474-3470cc18573c">

---

BTW, @liaowang11 could you change the title to `kitty: xxx` to comply with the [commit conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions)?

liaowang11 commented:
> Looks like this don't work.
> <img alt="image" width="800" src="https://private-user-images.githubusercontent.com/20698483/382221044-77d5a979-6f9a-4132-9526-7a3888948723.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA2ODU0NzUsIm5iZiI6MTczMDY4NTE3NSwicGF0aCI6Ii8yMDY5ODQ4My8zODIyMjEwNDQtNzdkNWE5NzktNmY5YS00MTMyLTk1MjYtN2EzODg4OTQ4NzIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA0VDAxNTI1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2YWY1MGEzNTdmM2QwZTgzOGRlYzBhZWQ1MWViMDgyNDY3MmExYjdjYTRjMTUxYzM5YjEzYzFhZmNkZjM2YjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jr96FcTg34iVmc-zDzfHj2FtTV7GAIZAMBC4x2szCwc">

Yes, This still does not work on and intel mac, I'm still trying to troubleshoot the culprit.

XYenon commented:
It looks like this is because the x86_64 version is unsigned and unnotarized.

https://forums.mozillazine.org/viewtopic.php?p=14987067#p14987067

XYenon commented:
> It looks like this is because the x86_64 version is unsigned and unnotarized.
> 
> [forums.mozillazine.org/viewtopic.php?p=14987067#p14987067](https://forums.mozillazine.org/viewtopic.php?p=14987067#p14987067)

It works fine after I added `autoSignDarwinBinariesHook`.

<img width="510" alt="image" src="https://github.com/user-attachments/assets/34c336d7-0093-4695-a6bc-8343e0bbae26">

```diff
diff --git a/pkgs/applications/terminal-emulators/kitty/default.nix b/pkgs/applications/terminal-emulators/kitty/default.nix
index 74365d300ec4..70bb01e9346f 100644
--- a/pkgs/applications/terminal-emulators/kitty/default.nix
+++ b/pkgs/applications/terminal-emulators/kitty/default.nix
@@ -29,6 +29,7 @@
 , buildGo123Module
 , nix-update-script
 , makeBinaryWrapper
+, autoSignDarwinBinariesHook
 }:
 
 with python3Packages;
@@ -90,6 +91,7 @@ buildPythonApplication rec {
   ] ++ lib.optionals stdenv.hostPlatform.isDarwin [
     imagemagick
     libicns  # For the png2icns tool.
+    autoSignDarwinBinariesHook
   ] ++ lib.optionals stdenv.hostPlatform.isLinux [
     wayland-scanner
   ];
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index c64bfa6a756a..43e8baf679e2 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -2825,6 +2825,7 @@ with pkgs;
 
   kitty = darwin.apple_sdk_11_0.callPackage ../applications/terminal-emulators/kitty {
     harfbuzz = harfbuzz.override { withCoreText = stdenv.hostPlatform.isDarwin; };
+    inherit (darwin) autoSignDarwinBinariesHook;
     inherit (darwin.apple_sdk_11_0) Libsystem;
     inherit (darwin.apple_sdk_11_0.frameworks)
       Cocoa
```

viniciuspalma commented:
Is this the same problem as "the application dock does not have permission to open (null)"?

I can't open the kitty app that I have installed via home manager, via the dock or via raycast after the update to Sequoia.

liaowang11 commented:
Reword the commit message to comply with the convention. I incorporated the sign diff into this commit; I'm not sure if this is the right way to go. If not, I can open a new pull request.

voidcontext commented:
I can confirm, this patch solves the issue of running kitty ("the application finder does not have permission to open (null)") on 15.1 Sequoia and `x86_64-darwin`. Thanks for the fix!

samueltardieu commented:
@liaowang11 Do the tests pass without disabling the sandbox on aarch64-darwin? See https://github.com/NixOS/nixpkgs/issues/353489.

liaowang11 commented:
> @liaowang11 Do the tests pass without disabling the sandbox on aarch64-darwin? See #353489.

No, build fails recently with sandbox on aarch64-darwin. Seems to be related to recently re-enabled tests in https://github.com/NixOS/nixpkgs/commit/4ff48958a2ecea05f8214350a85d1a56bc1152ef.

yzx9 commented:
Hello, any updates? As I understand it, #353489 shouldn’t block this PR, as @emilazy explained that building with the sandbox on is not always required on Darwin

Kernald commented:
This PR is also not causing the issue in sandbox - it's just addressing another one.

yzx9 commented:
@wegank Hello, could I ask for your attention on this PR?
